source/common/router/router.cc

```diff
@@ -602,6 +602,30 @@ void Filter::onUpstreamHeaders(const uint64_t response_code, Http::HeaderMapPtr&
     headers->addReferenceKey(Http::Headers::get().SetCookie, header_value);
   }
 
+  for (const auto& header_value : route_entry_->responseHeadersToAdd()) {
```

Comment:
Given the duplication here, and potentially more in the future, can we consolidate this at all? Either have a shared structure for all of the three levels, and shared methods? I think this is related to the vestigial request header code and header parser stuff that I mentioned.